File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case09_state/Case01_ValueState.scala
package com.yw.flink.example.scalacases.case09_state

import com.yw.flink.example.StationLog
import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.functions.KeyedProcessFunction
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.util.Collector

/**
  * Flink ValueState 状态测试
  * 案例：读取基站日志数据，统计每个主叫手机通话间隔时间，单位为毫秒
  * 001,186,187,busy,1000,10
  * 002,187,186,fail,2000,20
  * 003,186,188,busy,3000,30
  * 004,187,186,busy,4000,40
  * 005,189,187,busy,5000,50
  *
  */
object Case01_ValueState {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    //导入隐式转换
    import org.apache.flink.streaming.api.scala._
    val ds: DataStream[String] = env.socketTextStream("node5", 9999)
    val ds2: DataStream[StationLog] = ds.map(line => {
      val arr: Array[String] = line.split(",")
      StationLog(arr(0), arr(1), arr(2), arr(3), arr(4).toLong, arr(5).toLong)
    })

    ds2.keyBy(_.callOut).process(new KeyedProcessFunction[String, StationLog, String] {

      private var valueState: ValueState[Long] = _

      override def open(parameters: Configuration): Unit = {
        //定义状态描述符
        val descriptor = new ValueStateDescriptor[Long]("callTimeValueState", classOf[Long])
        valueState = getRuntimeContext.getState(descriptor)
      }


      //来一条数据处理一条
      override def processElement(stationLog: StationLog,
                                  context: KeyedProcessFunction[String, StationLog, String]#Context,
                                  collector: Collector[String]): Unit = {

        //获取状态
        val callTime: Long = valueState.value()
        if (callTime == 0) {
          //当前状态中没有值，第一通话，更新状态
          valueState.update(stationLog.callTime)
        } else {
          //计算当前通话时间与状态中的时间差值
          val intervalTime: Long = stationLog.callTime - callTime
          collector.collect("当前主叫号码：" + stationLog.callOut + s",两次通话间隔：${intervalTime}")
          //更新状态
          valueState.update(stationLog.callTime)
        }

      }
    }).print()

    env.execute()
  }

}

File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case09_state/Case01_ValueState.java
package com.yw.flink.example.javacases.case09_state;

import com.yw.flink.example.StationLog;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.java.functions.KeySelector;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.util.Collector;

/**
 * Flink ValueState 状态测试
 * 案例：读取基站日志数据，统计每个主叫手机通话间隔时间，单位为毫秒
 * 001,186,187,busy,1000,10
 * 002,187,186,fail,2000,20
 * 003,186,188,busy,3000,30
 * 004,187,186,busy,4000,40
 * 005,189,187,busy,5000,50
 */
public class Case01_ValueState {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        //读取socket数据
        DataStreamSource<String> ds = env.socketTextStream("nc_server", 9999);
        //转换数据 ds -> StationLog
        SingleOutputStreamOperator<StationLog> stationLogDS = ds.map(new MapFunction<String, StationLog>() {
            @Override
            public StationLog map(String line) throws Exception {
                String[] split = line.split(",");
                return new StationLog(split[0], split[1], split[2], split[3], Long.valueOf(split[4]), Long.valueOf(split[5]));
            }
        });

        //对 stationLogDS 进行keyby ,key选择的是主叫
        KeyedStream<StationLog, String> keyedStream = stationLogDS.keyBy((KeySelector<StationLog, String>) stationLog -> stationLog.callOut);

        //状态编程，统计每个主叫号码两次通话间隔时长
        SingleOutputStreamOperator<String> result = keyedStream.process(new KeyedProcessFunction<String, StationLog, String>() {

            private ValueState<Long> callTimeValueState;

            //状态注册与获取
            @Override
            public void open(Configuration parameters) throws Exception {
                //1.定义状态描述器
                ValueStateDescriptor<Long> descriptor = new ValueStateDescriptor<>("callTimeValueState", Long.class);
                //2.获取状态
                callTimeValueState = getRuntimeContext().getState(descriptor);
            }

            //来一条数据处理一条数据
            @Override
            public void processElement(StationLog stationLog, KeyedProcessFunction<String, StationLog, String>.Context context, Collector<String> collector) throws Exception {
                //3.获取状态值
                Long callTime = callTimeValueState.value();
                if (callTime == null) {
                    //说明状态为空，这里该主叫的第一次通话，将通话时间直接存储在状态中
                    callTimeValueState.update(stationLog.callTime);
                } else {
                    //状态中有值，计算两次通话时长
                    long intervalTime = stationLog.callTime - callTime;
                    collector.collect("主叫 ：" + stationLog.callOut + ",最近两次通话间隔时长：" + intervalTime);
                    //4.更新状态
                    callTimeValueState.update(stationLog.callTime);
                }
            }
        });

        result.print();
        env.execute();
    }
}
